binary_tree.c

```diff
@@ -0,0 +1,410 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include "node.h"
+#include "bfs_queue.h"
+
+typedef struct {
+    struct Node* root;
+} Tree;
+
+
+Tree tree_create(); // returns empty binary tree
+// void tree_free(Tree* self); // deallocates tree
+
+struct Node* tree_create_node(int);
+void tree_delete_node(Tree*, struct Node*);
+
+int tree_get_node_val(struct Node*);
+void tree_set_node_val(struct Node*, int);
+
+struct Node* tree_get_left_child(struct Node*);
+struct Node* tree_get_right_child(struct Node*);
+
+void tree_insert_left(struct Node*, int);
+void tree_insert_right(struct Node*, int);
+
+//actually not the part of the interface, part of learning
+void tree_inorder_traverse(struct Node*);
+void tree_preorder_traverse(struct Node*);
+void tree_postorder_traverse(struct Node*);
+
+struct Node* tree_search_parent(Tree*, struct Node*);
+struct Node* tree_search_last_right_vertice(struct Node*);
+struct Node* tree_dfs_search(struct Node*, int); //inorder traversal
+struct Node* tree_bfs_search(struct Node*, int);
+
+int tree_get_node_val(struct Node* self) {
+    if (self->data) {
+        return self->data;
+    }
+    else {
+        printf("Node has no data.\n");
+        exit(EXIT_FAILURE);
+    }
+}
+
+void tree_set_node_val(struct Node* self, int value) {
+    if (self) {
+        self->data = value;
+    }
+    else {
+        printf("Node pointer is NULL\n");
+        exit(EXIT_FAILURE);
+    }
+}
+
+struct Node* tree_get_left_child(struct Node* self) {
+    if (self->l_node) {
+        return self->l_node;
+    }
+    else {
+        return NULL;
+    }
+}
+
+struct Node* tree_get_right_child(struct Node* self) {
+    if (self->r_node) {
+        return self->r_node;
+    }
+    else {
+        return NULL;
+    }
+}
+
+void tree_insert_left(struct Node* self, int value) {
+    struct Node* new_node = tree_create_node(value);
+
+    if(self->l_node) {
+        new_node->l_node=self->l_node;
+    }
+    self->l_node=new_node;
+}
+
+void tree_insert_right(struct Node* self, int value) {
+    struct Node* new_node = tree_create_node(value);
+
+    if(self->r_node) {
+        new_node->r_node=self->r_node;
+    }
+
+    self->r_node=new_node;
+}
+
+struct Node* tree_bfs_search(struct Node* self, int value) {
+    struct Node *n = NULL;
+    enQueue(self);
+
+    while((n = deQueue()) != NULL) {
+
+        if(n->data==value) {
+            return n;
+        }
+        if(n->l_node) {
+            enQueue(n->l_node);
+        }
+        if(n->r_node) {
+            enQueue(n->r_node);
+        }
+    }
+    return n;
+}
+
+struct Node* tree_dfs_search(struct Node* self, int value) {
```

Comment:
well, DFS is a traverse method, you don't need to search with DFS in binary tree, since it's already sorted. There's only one search algorithm possible, it's comparing if current node is less, greater or equal to the vlue being searched

Reply:
Shall I add sort method then?

Reply:
No. The tree always remains sorted